### ISSUE TITLE
Feature request: part of speech separation

### ISSUE THREAD

**Issue by chopinesque** [BoboTiG/ebook-reader-dict#1149]
Currently, all definitions for each entry are in one list regardless of part of speech (verb, noun, etc). It would be good if at least there was an option to have them grouped by part of speech for easier accessing of information.

**Comment by lasconic** [comment]
A bit like https://github.com/BoboTiG/ebook-reader-dict/issues/1104, it would be possible to do that if we could store several (word + definition) for a given word. Right now, `Words` is a `dict(str, Word)`  and `Word`  a list of etymologies, a single part of speech, pronunciation and a list of definition. Somehow we would need to have Words = dict[str, list(Word)]...
Also we would need to match the etymology to the right part of speech...

**Comment by chopinesque** [comment]
I think that etymology should/would appear at the top no matter what. Mind you, I think etymology too could benefit from an option not to include it in the final file as most people are not really that interested about it.

Of course, you are the expert on the coding aspect and the feasibility thereof, but as a daily professional dictionary user I can see great benefits in part of speech (POS) sense grouping. At the end of the day, they are already grouped as such in Wiktionaries. 

Another option would be to have a prefix in each sense with the POS (as to avoid using different headers per POS) and have them sorted on that prefix. Not sure if that makes any sense or whether it would be easier programmatically.

**Comment by Moonbase59** [comment]
I absolutely agree with @chopinesque.

Fortunately, I have a few dictionaries installed, so I append an HTML output showing how others do the part-of-speech separation for the word "test".

Of course I prefer mine ("Wiktionary (De-De)"), hee hee, and ours (this project’s) is under "Wiktionary EN-EN".

[test.html.zip](https://github.com/BoboTiG/ebook-reader-dict/files/7969778/test.html.zip)

**Comment by victornove** [comment]
Hi, 
My two cents on the topic. In my first attempt, I added the part of speech to each definition of the word. It seemed to be the simplest and most flexible way to do it. For the languages I checked the POS was usually in the section title, but for german and russian I just injected it into the section title from previous/higher sections in the hierarchy. This only solves half of your problem though, since you need to render it in a dictionary afterwards.

**Comment by MolotovCherry** [comment]
I want to say, this project is awesome. I absolutely love it. But regarding what the other person said about etymology, I have to say I also agree. It is getting in the way of my reading, and I would just like to see definitions with no etymology or pronunciation

**Comment by BoboTiG** [comment]
Then, let's generate additionnal dictionnaries without etymology. I'll try something to see how it fits.

**Comment by BoboTiG** [comment]
You can try etymology-free dictionaries right now ;)

**Comment by MolotovCherry** [comment]
> You can try etymology-free dictionaries right now ;)

Really appreciate it! Thanks a lot!

**Comment by lasconic** [comment]
This FR is one of the oldest. I would like to revisit it. I read the thread again and I just realized why @BoboTiG talked about a new template engine. My plan was to use the Kobo feature that let add several times the same word in the dictionary and will concatenate the HTML at the time of display. @BoboTiG suggestion was to construct an HTML document with all the information of a word.

One of the main issue in both cases is to have the right internal representation. If we only take English and French, words with different part of speech (PoS) are rendered differently in the wiktionary.

In english see https://en.wiktionary.org/wiki/lead
```
Etymology 1
   Etymology 1 content
   Pronunciation
   Noun - PoS1 (metadata we don't consider right now such as "countable")
       Definitions
   Verb - PoS2
       Definitions

Etymology 2
    Etymology 2 content
   Pronunciation
   Verb - PoS1
        Definitions
   Noun - PoS2
        Definitions
   Adjective - PoS3
        Definitions

Etymology 3
    Verb - PoS1
        Definitions
```

In french, see for example Moule :  https://fr.wiktionary.org/wiki/moule (or https://fr.wiktionary.org/wiki/voile  https://fr.wiktionary.org/wiki/poule  https://fr.wiktionary.org/wiki/livre)

```
Etymology
   etymology Noun 1
   etymology Noun 2
Noun Pos1
   Pronunciation
   genre
   definitions
Noun Pos1
   Pronunciation
   genre
   definitions
```

Some examples are even more complex see : https://fr.wiktionary.org/wiki/juste

So far we tried to have a consistent output for all the languages. Which output should we use here ? Do we still aim for consistency ? It will probably influence the way we store the data internally.

**Comment by BoboTiG** [comment]
It is relevant for English and French, but I guess every other languages will need special handling too.

**Comment by BoboTiG** [comment]
Side personal note: I like the way our dictionaries are quite concise, and most of the time it is not really needed to scroll to display definitions. Adding POS means adding more blank lines, and possibily multiple etymologies too. So from a compact design we would move to something more verbose.

it's just a note, not a no-go or something else.

**Comment by chopinesque** [comment]
The current conciseness could be maintained with the following option. Conciseness does not work well if one does not know where to look for the verb or noun meanings as it takes longer to find what one is looking for.
> Another option would be to have a prefix in each sense with the POS (as to avoid using different headers per POS) and have them sorted on that prefix.

**Comment by BoboTiG** [comment]
Ah yes, I was thinking about this option too. It could work 👍

**Comment by chopinesque** [comment]
I guess technically this is much more straightforward, is this correct?

**Comment by BoboTiG** [comment]
Not really haha. Using this option, we "only" prevent us from creating a template engine. The parsing mechanism still needs to be rewritten though.

**Comment by BoboTiG** [comment]
Working on a solution 🚧

**Comment by BoboTiG** [comment]
A first move in the right direction:
- prepend the POS in small characters
- definitions are sorted by POS
- POS are kept as-is (no short prefix) because I do not know what is the best, and we have so many langs to handle that it may become tricky, but maybe still doable with some help)

Preview, what do you think?

<details>

![Image](https://github.com/user-attachments/assets/04ccdd65-2d28-48b1-934a-3ddcbeb8eb0b)
</details>

<details>

![Image](https://github.com/user-attachments/assets/9de80555-56e9-4928-b506-fbdce606d1af)
</details>

**Comment by chopinesque** [comment]
> prepend the POS in small characters

Maybe add some per POS colour (ie x color for verb, y color for noun) or other visual clue to make it stand out (so that one can easily navigate visually to the different POS).
When sorting per POS, retain Wiktionary sorting?

The alphabetic sorting of which POS category comes first may not be linguistically intuitive (i.e. verbs and nouns should precede adjectives).

Not sure if this is related, but I see no space after comma in some instances, ie **by,buy,bye**

**Comment by BoboTiG** [comment]
> Not sure if this is related, but I see no space after comma in some instances, ie **by,buy,bye**

This is actually an issue when rendering "form of" templates. It'll be tackled later.

**Comment by BoboTiG** [comment]
> When sorting per POS, retain Wiktionary sorting?

I went with alphabetical sort, as you proposed in https://github.com/BoboTiG/ebook-reader-dict/issues/1149#issuecomment-1010896158 (but maybe I misunderstood). We need sorting at some point, because we would have mixed POS then.

**Comment by BoboTiG** [comment]
> Maybe add some per POS colour (ie x color for verb, y color for noun) or other visual clue to make it stand out (so that one can easily navigate visually to the different POS).

I would like to start "light", and then iterate to improve.

First, we need to support all locales, and when I see https://github.com/BoboTiG/ebook-reader-dict/blob/ba66b968914043c526b4c2d86dea5c1b7e48c82f/wikidict/lang/da/__init__.py#L29-L95 with long POS names, finally a short version would be better.

**Comment by chopinesque** [comment]
> > When sorting per POS, retain Wiktionary sorting?
> 
> I went with alphabetical sort, as you proposed in [#1149 (comment)](https://github.com/BoboTiG/ebook-reader-dict/issues/1149#issuecomment-1010896158) (but maybe I misunderstood). We need sorting at some point, because we would have mixed POS then.

Yes, this is a refinement, a custom meta-level of order, after per POS sorting. For example (note the numbering restart):

1. [noun] Sense 1
2. [noun] Sense 2
3. [noun] Sense 3

**[Some space here or other visual cue?]**

1. [verb] Sense 1
2. [verb] Sense 2
3. [verb] Sense 3

Of course you can start light and build gradually on refinements.

**Comment by BoboTiG** [comment]
Good idea.

But the more I think about it, the more it would better to tackle the subject entirely and split by POS directly. Because here, I'll anyway have to change the logic.

I'll try something, and post an updated preview.

**Comment by lasconic** [comment]
I agree. If you do put some space (or visual cue) between POS, then why not the POS name

**Comment by BoboTiG** [comment]
How is it?

<details>

![Image](https://github.com/user-attachments/assets/9cf6bdeb-cdd9-429d-9603-d2165630f9f2)
</details>

<details>

![Image](https://github.com/user-attachments/assets/96cb6f23-d9ca-42e4-bd8d-5eb2754cc3c1)
</details>

<details>

![Image](https://github.com/user-attachments/assets/678d676a-3bbc-4a1f-ab67-1075232464b5)
</details>

**Comment by lasconic** [comment]
Wow. It's great ! It works for french as well ?

**Comment by chopinesque** [comment]
Looks quite good to me. Maybe make POS bold?

**Comment by BoboTiG** [comment]
> Wow. It's great ! It works for french as well ?

For now, it works for CA, DA, EN and FR.
I am still working on others: ultimately, all supported locales will have POS.

**Comment by BoboTiG** [comment]
> Looks quite good to me. Maybe make POS bold?

I thought to do that. I'll try, and see if the Kindle is happy with that (I recall there is a thing with the first bold text occurence in the definition, I'll try and see what happens).

**Comment by BoboTiG** [comment]
With the POS in bold, it is way better 👍 

<details>

![Image](https://github.com/user-attachments/assets/85c1ea2f-2da8-4052-aa09-905660fb69fd)
</details>

**Comment by chopinesque** [comment]
Yes, excellent work, congratulations!
I wonder how it would function if there are multiple sections with the same POS (and possibly different etymologies), will they be merged on per POS basis?

**Comment by BoboTiG** [comment]
> I wonder how it would function if there are multiple sections with the same POS (and possibly different etymologies), will they be merged on per POS basis?

Yes, they will be merged together under their POS.

Etymologies are different beasts too. I do not touch them right now.

**Comment by chopinesque** [comment]
Etymologies would be very low on my priority list anyway due to the fact that very few people need to check them out.
I think that the new format, once applied to other languages too (still, the main impact will be for English due to the fact that lemmas often have multiple POS), will be a groundbreaking change in this project, and it will allow for the optional inclusion of other sections like for example expressions https://github.com/BoboTiG/ebook-reader-dict/issues/2438#issuecomment-2905483025, thus further enriching the dictionaries in a clear and unambiguous way.

**Comment by BoboTiG** [comment]
@chopinesque what form should we prefer between "Επίθημα" and "Επίρρημα" for the POS? I will merge one into the other to prevent duplicates.

**Comment by chopinesque** [comment]
They are 2 different categories, one is **suffix**, the other **adverb**.

**Comment by BoboTiG** [comment]
> but for german and russian I just injected it into the section title from previous/higher sections in the hierarchy.

@victornove do you remember some code for this? I would be interested in how you did for German (cf #2468) :)
